File: Week 07/id_002/sort.js
// 1. 冒泡
function bubblesort(arr) {
    var len = arr.length;
    for (let i = 0; i < len; i ++) {
        for (j = 0; j < len - i; j ++) {
            if (arr[j] > arr[j + 1]) {
                [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
            }
        }
    }

    return arr;
}

// 2. 插入排序
function insertionSort(arr) {
    let len = arr.length;
   
    for (let i = 0; i < len - 1; i ++) {
        let cur = i;

        while(cur - 1 >= 0 && arr[cur - 1] > arr[cur]) {
            [arr[cur - 1], arr[cur]] = [arr[cur], arr[cur - 1]];
            cur --;
        }
    }

    return arr;
}

// 3. 选择排序
function selectionSort(arr) {
    let len = arr.length;
    let minIdx = 0;

    for (let i = 0; i < len; i ++) {
        minIdx = i;
        for (let j = i + 1; j < len;  j++) {
            if (arr[j] < arr[minIdx]) minIdx = j;
        }

        [arr[i], arr[minIdx]] = [arr[minIdx], arr[i]];
    }
}

// 4. 快速排序
function quickSort(arr, begin = 0, end = arr.length - 1) {
    if (begin >= end) return;

    let pivot = partition(begin, end);
    quickSort(arr, begin, pivot - 1);
    quickSort(arr, pivot + 1, end);
    
    function partition(begin, end) {
        let pivot = end;
        let counter = begin;

        for (let i = begin; i < end; i ++) {
            if (arr[i] < arr[pivot]) {
                let temp = arr[counter]; arr[counter] = arr[i]; arr[i] = temp;

                counter ++;
            }
        }

        let temp = arr[pivot]; arr[pivot] = arr[counter]; arr[counter] = temp;

        return counter;
    }
}

// 5. 归并排序
function mergeSort(array, left = 0, right = array.length - 1) {
    if (left >= right) return;

    let mid = (left + right) >> 1;

    mergeSort(array, left, mid);
    mergeSort(array, mid + 1, right);
    merge(array, left, mid, right);

    function merge(array, left, mid, right) {
        let i = left;
        let j = mid + 1;
        let k = 0;
        let tmp = [];
    
        while(i <= mid && j <= right) {
            tmp[k++] = array[i] <= array[j] ? array[i++] : array[j++];
        }
    
        while(i <= mid) temp[k++] = array[i++];
        while(j <= right) temp[k++] = array[j++];
    
        for(let p = 0; p < temp.length; p ++) {
            array[left + p] = temp[p];
        }
    }
}

// 6. 堆排序
 function heapSort(arr) {
    let len = arr.length;
    
    // 堆化
    for (let i = (len >> 1) -1; i >= 0; i ++) {
        heapify(arr, len, i);
    }

    // 取出顶部元素放入 数组尾部， 调整堆
    for (let i = len - 1; i >= 0; i ++) {
        let temp = arr[0]; arr[0] = arr[i]; arr[i] = temp;
        heapify(arr, i, 0);
    }
}

function heapify(arr, length, i) {
    let left = 2 * i + 1;
    let right = 2 * i + 2;
    let largest = i;

    if (left < length && arr[left] > arr[largest]) largest = left;

    if (right < length && arr[right] > arr[largest]) largest = right;

    if (largest != i) {
        let temp = arr[i];  arr[i] = arr[largest]; arr[largest] = temp;
        heapify(arr, length, largest);
    }
}

File: Week 07/id_372/sort/sort.go
package sort

func selectionSort(nums []int) []int {
	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[j] < nums[i] {
				nums[i], nums[j] = nums[j], nums[i]
			}
		}
	}

	return nums
}

func bubbleSort(nums []int) []int {
	for i := 0; i < len(nums); i++ {
		for j := 0; j < len(nums)-1; j++ {
			if nums[j] > nums[j+1] {
				nums[j], nums[j+1] = nums[j+1], nums[j]
			}
		}
	}

	return nums
}

func insertionSort(nums []int) []int {
	for i := 0; i < len(nums)-1; i++ {
		for j := i + 1; j > 0; j-- {
			if nums[j] < nums[j-1] {
				nums[j], nums[j-1] = nums[j-1], nums[j]
			}
		}
	}

	return nums
}

func quickSort(nums []int, start, end int) {
	if start >= end {
		return
	}

	pivot := partition(nums, start, end)
	quickSort(nums, start, pivot-1)
	quickSort(nums, pivot+1, end)
}

func partition(nums []int, start, end int) int {
	pivot, counter := end, start
	for i := start; i < end; i++ {
		if nums[i] < nums[pivot] {
			nums[i], nums[counter] = nums[counter], nums[i]
			counter++
		}
	}
	nums[pivot], nums[counter] = nums[counter], nums[pivot]

	return counter
}

func mergeSort(nums []int, left, right int) {
	if left >= right {
		return
	}
	mid := left + (right-left)/2

	mergeSort(nums, left, mid)
	mergeSort(nums, mid+1, right)
	merge(nums, left, mid, right)
}

func merge(nums []int, left, mid, right int) {
	tmp := make([]int, right-left+1)
	i, j, k := left, mid+1, 0

	for i <= mid && j <= right {
		if nums[i] <= nums[j] {
			tmp[k] = nums[i]
			i++
			k++
		} else {
			tmp[k] = nums[j]
			j++
			k++
		}
	}

	for i <= mid {
		tmp[k] = nums[i]
		i++
		k++
	}

	for j <= right {
		tmp[k] = nums[j]
		j++
		k++
	}

	for p := 0; p < len(tmp); p++ {
		nums[left+p] = tmp[p]
	}
}
